### PR TITLE
Implement high contrast mode

### PR DESCRIPTION
**Pull request by sawanm9000**
Fixes issue #13109.

https://user-images.githubusercontent.com/25604837/114038925-d065f000-9871-11eb-8dc0-bb628abfcdc7.mp4

### CODE REVIEW

File: web/viewer.css [hunk 1]
```diff
@@ -129,18 +135,42 @@
...
     --doorhanger-separator-color: rgba(92, 92, 97, 1);
     --overlay-button-bg-color: rgba(92, 92, 97, 1);
-    --overlay-button-hover-color: rgba(115, 115, 115, 1);
+    --overlay-button-hover-bg-color: rgba(115, 115, 115, 1);
 
     /* This image is used in <input> elements, which unfortunately means that
-     * the `mask-image` approach used with all of the other images doesn't work
-     * here; hence why we still have two versions of this particular image. */
+    * the `mask-image` approach used with all of the other images doesn't work
+    * here; hence why we still have two versions of this particular image. */
```
Comment (Snuffleupagus) [inline web/viewer.css:146]: Here, and elsewhere, please don't re-format existing comments.

File: web/viewer.css [hunk 2]
```diff
@@ -129,18 +135,42 @@
...
     --loading-icon: url(images/loading-dark.svg);
   }
 }
 
+@media screen and (forced-colors: active) {
+    :root {
+      /* --dropdown-btn-bg-color: red !important; */
+      /* --main-color: rgba(255, 255, 255, 1); */
```
Comment (Snuffleupagus) [inline web/viewer.css:154]: Please remove, rather than comment out, unused rules.

File: web/viewer.css [hunk 3]
```diff
@@ -1758,24 +1829,23 @@ html[dir="rtl"] #documentPropertiesOverlay .row > * {
...
     flex-direction: column;
     justify-content: center;
     align-items: center;
   }
   #printContainer canvas,
   #printContainer img {
-    /* The intrinsic canvas / image size will make sure that we fit the page. */
-    max-width: 100%;
-    max-height: 100%;
-
```
Comment (Snuffleupagus) [inline web/viewer.css:1778]: These changes look first of all unrelated to *this* patch, and secondly they're basically reverting a bunch of recent printing fixes; please undo these changes.

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

File: web/viewer.css [hunk 4]
```diff
@@ -129,18 +135,42 @@
...
+      --toggled-btn-color: HighlightText;
+      --toggled-btn-bg-color: Highlight;
+      --doorhanger-hover-color: ButtonFace;
+      --doorhanger-border-color-whcm: 1px solid ButtonText;
+      --doorhanger-hat-opacity-whcm: 0;
+      --overlay-button-border: 1px solid Highlight;
+      --overlay-button-hover-bg-color: Highlight;
+      --overlay-button-hover-color: ButtonFace;
+      --field-border-color: ButtonText;
+
```
Comment (Snuffleupagus) [inline web/viewer.css:170]: Nit: unnecessary newline, please remove.

File: web/viewer.css [hunk 5]
```diff
@@ -129,18 +135,42 @@
...
+      --button-hover-color: Highlight;
+      --doorhanger-hover-bg-color: Highlight;
+      --toolbar-icon-opacity: 1;
+      --toolbar-icon-bg-color: ButtonText;
+      --toolbar-icon-hover-bg-color: ButtonFace;
+      --toggled-btn-color: HighlightText;
+      --toggled-btn-bg-color: Highlight;
+      --doorhanger-hover-color: ButtonFace;
+      --doorhanger-border-color-whcm: 1px solid ButtonText;
+      --doorhanger-hat-opacity-whcm: 0;
```
Comment (Snuffleupagus) [inline web/viewer.css:165]: What does "hat" mean, please use a more descriptive name?

File: web/viewer.css [hunk 6]
```diff
@@ -722,18 +757,30 @@ html[dir="rtl"] .splitToolbarButton > .toolbarButton {
...
 }
 .overlayButton {
   background-color: var(--overlay-button-bg-color);
+  border-radius: 2px;
+  color: var(--main-color);
+  font-size: 12px;
+  line-height: 14px;
+  user-select: none;
+  cursor: default;
+  box-sizing: border-box;
```
Comment (Snuffleupagus) [inline web/viewer.css:773]: Is it really necessary to duplicate these rules here, i.e. why [remove this line](https://github.com/mozilla/pdf.js/pull/13203/files#diff-d1047e1376717fbe7c554bb385cd48b2dcbbb135e02884be3db97ad072ba741eL800)?

File: web/viewer.css [hunk 7]
```diff
@@ -1262,9 +1317,13 @@ html[dir="rtl"] .toolbarField[type="checkbox"] {
 
 html[dir="ltr"] #numPages.toolbarLabel {
   padding-left: 2px;
+  padding-top: 7px;
+  margin: 2px;
```
Comment (Snuffleupagus) [inline web/viewer.css:1321]: This moves the `#numPages` element, please revert this line!

File: web/viewer.css [hunk 8]
```diff
@@ -1262,9 +1317,13 @@ html[dir="rtl"] .toolbarField[type="checkbox"] {
...
 }
 html[dir="rtl"] #numPages.toolbarLabel {
   padding-right: 2px;
+  padding-top: 7px;
+  margin: 2px;
```
Comment (Snuffleupagus) [inline web/viewer.css:1326]: This moves the `#numPages` element, please revert this change!

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

File: web/viewer.css [hunk 9]
```diff
@@ -1249,8 +1305,7 @@ html[dir="rtl"] .toolbarField[type="checkbox"] {
 .toolbarLabel {
   min-width: 16px;
   padding: 6px;
-  margin: 2px;
-  border: 1px solid rgba(0, 0, 0, 0);
```
Comment (Snuffleupagus) [inline web/viewer.css:1253]: This change moves/shrinks the `#findResultsCount` label, please avoid that.

Comment (utopianknight) [reply web/viewer.css:1253]: The only way to remove the border in high contrast mode is by removing this line.
Screenshot 2021-03-16 123114

Comment (Snuffleupagus) [reply web/viewer.css:1253]: Hmm, what about something like this then (the diff is against `master`, hopefully that's not a problem)?
```diff --git a/web/viewer.css b/web/viewer.css
index 442f57580..4a45149b8 100644
--- a/web/viewer.css
+++ b/web/viewer.css
@@ -660,7 +660,7 @@ html[dir="ltr"] .doorHangerRight:before {
   background-color: rgba(217, 217, 217, 1);
   color: rgba(82, 82, 82, 1);
   text-align: center;
-  padding: 3px 4px;
+  padding: 4px 5px;
   margin: 5px;
 }
 
@@ -1248,9 +1248,8 @@ html[dir="rtl"] .toolbarField[type="checkbox"] {
 
 .toolbarLabel {
   min-width: 16px;
-  padding: 6px;
+  padding: 7px;
   margin: 2px;
-  border: 1px solid rgba(0, 0, 0, 0);
   border-radius: 2px;
   color: var(--main-color);
   font-size: 12px;
@@ -1261,10 +1260,10 @@ html[dir="rtl"] .toolbarField[type="checkbox"] {
 }
 
 html[dir="ltr"] #numPages.toolbarLabel {
-  padding-left: 2px;
+  padding-left: 3px;
 }
 html[dir="rtl"] #numPages.toolbarLabel {
-  padding-right: 2px;
+  padding-right: 3px;
 }
 
 #thumbnailView {
```

File: web/viewer.css [hunk 10]
```diff
@@ -878,16 +923,20 @@ html[dir="ltr"] .dropdownToolbarButton > select {
...
-.dropdownToolbarButton > select:hover {
+.dropdownToolbarButton > select:hover,
+.dropdownToolbarButton > select:focus {
   background-color: var(--button-hover-color);
+  color: var(--toggled-btn-color);
 }
 
-.dropdownToolbarButton > select:focus {
-  background-color: var(--button-hover-color);
+.dropdownToolbarButton > select:active {
```
Comment (Snuffleupagus) [inline web/viewer.css:932]: Given than none of the "regular" buttons use any `select:active` rules, this feels a little jarring then clicking on the zoom-dropdown; do we *really* need to add this rule?

Comment (utopianknight) [reply web/viewer.css:932]: I wanted to make it behave like the 'toggle sidebar', 'find button' and the 'tool' menu button but it didn't work well so I'm reverting it.

**Comment by brendandahl** [comment]
@Snuffleupagus  or @timvandermeij is this good to go? I'm hoping to do a mozilla-central update soon.

**Comment by Snuffleupagus** [comment]
> @Snuffleupagus or @timvandermeij is this good to go? I'm hoping to do a mozilla-central update soon.

*Please note:* I've not really tested the `forced-colors` part of the patch, and instead focused on ensuring that this doesn't break/regress the "normal" light/dark themes.
The latest preview addresses all my previous comments, and as far as I'm concerned it's good to to go now :-)

@brendandahl As far as I understand, you need to manually set `layout.css.forced-colors.enabled = true` in about:config in order for this patch to actually have an effect!?
Is that preference going to be toggled *on* for all Firefox users? If not, can we special-case it (on the mozilla-central side) to apply to code running at `resource://pdf.js/web/viewer.html`?

**Comment by brendandahl** [comment]
> Is that preference going to be toggled _on_ for all Firefox users? If not, can we special-case it (on the mozilla-central side) to apply to code running at `resource://pdf.js/web/viewer.html`?

It will soon on by default https://groups.google.com/g/mozilla.dev.platform/c/zwDaj0JMYjs
